File: DSA/String/0020_Valid_Parentheses/cpp/solution_2.cpp
// 0020. Valid Parentheses
// Difficulty : Easy
// Tags       : string, stack
// Solved on  : 2020-12-17
// Attempt    : #2
// LeetCode   : https://leetcode.com/problems/valid-parentheses/

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(char c : s)
        {
            if(c=='(' || c=='{' || c=='[')
            {
                st.push(c);
            }
            else if(c==')' || c=='}' || c==']')
            {   if(st.size()<=0)
                    return false;
                if(c==')'){
                    char ch=st.top();
                    if(ch!='(') return false;
                    st.pop();
                }
                else if(c=='}')
                {
                    char ch=st.top();
                    if(ch!='{') return false;
                     st.pop();
                }
                else if(c==']')
                {
                    char ch=st.top();
                    if(ch!='[') return false;
                     st.pop();
                }
            }
            
            
           
        }
        if(st.size()>0) return false;
        return true;
    }
};

File: DSA/Math/0062_Unique_Paths/python3/solution_2.py
# 0062. Unique Paths
# Difficulty : Medium
# Tags       : math, dynamic-programming, combinatorics
# Solved on  : 2026-04-30
# Attempt    : #2
# LeetCode   : https://leetcode.com/problems/unique-paths/

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        return self.upath(0,0,m,n);
    @lru_cache(None)
    def upath(self,i:int,j:int,m:int,n:int):
        if i >=m or j >=n:
            return 0  
        if i == m-1 and j == n-1:
            return 1
        return self.upath(i+1,j,m,n) + self.upath(i,j+1,m,n)